### DISCUSSION TITLE
Per-yield type constraint between yield and send

### DISCUSSION THREAD

**Discussion by elbaro** [microsoft/pyright#7633] [Q&A] [answered]
The return type of a generator, `Generator[yield, send, return]`, is useful to annnotate when all `yield` statements have the same yield and send types. Can this be extended to annotate per-yield typing constraints such as 'When yielding a value of type T, the response of type T should be sent back'?

```py

from typing import Generator

def g() -> Generator[?, ?, None]:
  send1 = yield 123  # send1 is int
  send2 = yield 'abc' # send2 is str
```

A similar question is typing `list[tuple[T, Optional[T]]]` where T can be different for each item in the list.

**Comment by erictraut** [comment, marked as answer]
No, there's no way to provide per-yield types. You can create a union type to specify that the yield and send types can be _either_ `int` or `str` values.

`Generator[int | str, int | str, None]`

Similarly, there's no way to specify a list where different items in the list have different types. Lists elements are homogenous. This makes sense because lists are mutable, and elements can be replaced or moved within the list. Tuples, by comparison, are immutable, and their element types can be specified individually.
